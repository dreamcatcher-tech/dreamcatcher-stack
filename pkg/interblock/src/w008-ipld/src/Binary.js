import assert from 'assert-fast'
import { CID } from 'multiformats/cid'
import { encode, Block } from 'multiformats/block'
import { IpldInterface } from './IpldInterface'
import * as codec from 'multiformats/codecs/raw'
import { sha256 as hasher } from 'multiformats/hashes/sha2'

export class Binary extends IpldInterface {
  // TODO use the unixFS API so these objects are browseable
  #ipldBlock
  static async create(value) {
    assert(value instanceof Uint8Array)
    // TODO move to factory
    const ipldBlock = await encode({ value, codec, hasher })
    assert.strictEqual(ipldBlock.value, value)
    return this.#create(ipldBlock)
  }
  static #create(ipldBlock) {
    assert(ipldBlock instanceof Block, `#create requires a Block`)
    assert.strictEqual(ipldBlock.value, ipldBlock.bytes)
    const instance = new this()
    instance.#ipldBlock = ipldBlock
    instance.deepFreeze()
    IpldInterface.deepFreeze(ipldBlock)
    return instance
  }
  static async uncrush(rootCid, resolver, options) {
    assert(rootCid instanceof CID, `rootCid must be a CID, got ${rootCid}`)
    assert(typeof resolver === 'function', `resolver must be a function`)
    const ipldBlock = await resolver(rootCid)
    return this.#create(ipldBlock)
  }
  isModified() {
    // TODO verify that the Uint8Array cannot be altered
    return false
  }
  get ipldBlock() {
    return this.#ipldBlock
  }
  crush() {
    return this
  }
  getDiffBlocks(from) {
    if (this === from) {
      return new Map()
    }
    const blocks = new Map()
    blocks.set(this.ipldBlock.cid.toString(), this.ipldBlock)
    return blocks
  }
}
