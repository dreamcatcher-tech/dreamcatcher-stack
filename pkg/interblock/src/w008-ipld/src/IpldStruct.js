import assert from 'assert-fast'
import { CID } from 'multiformats/cid'
import { Block } from 'multiformats/block'
import { encode } from './IpldUtils'
import { IpldInterface } from './IpldInterface'

export class IpldStruct extends IpldInterface {
  #ipldBlock
  #previous
  #crushed
  #isPreCrushed = false
  static clone(map) {
    assert.strictEqual(typeof map, 'object')
    const instance = new this()
    Object.assign(instance, map)
    return instance
  }
  clone() {
    const next = new this.constructor()
    Object.assign(next, this)
    next.#ipldBlock = this.#ipldBlock
    next.#previous = this.#previous
    next.#crushed = this.#crushed
    next.#isPreCrushed = this.#isPreCrushed
    return next
  }
  isModified() {
    return this.#ipldBlock === undefined
  }
  get ipldBlock() {
    if (this.isModified()) {
      throw new Error('Instance must be crushed before block is encoded')
    }
    assert(this.#ipldBlock instanceof Block)
    return this.#ipldBlock
  }
  get crushedSize() {
    return this.ipldBlock.bytes.length
  }
  get cid() {
    const { cid } = this.ipldBlock
    return cid
  }
  get currentCrush() {
    return this.#crushed
  }
  async crush(resolver) {
    if (!this.isModified()) {
      assert(this.#ipldBlock)
      assert(this === this.#crushed)
      if (this !== this.#previous) {
        const next = this.clone()
        next.#previous = next
        next.#crushed = next
        return next
      }
      return this
    }
    if (this.#isPreCrushed) {
      // throw new Error('TESTING Crush has already been called')
    }
    this.#isPreCrushed = true
    const crushed = new this.constructor()
    Object.assign(crushed, this)
    const dagTree = { ...this }
    for (const key in crushed) {
      const slice = crushed[key]
      if (slice instanceof IpldInterface) {
        crushed[key] = await slice.crush(resolver)
        if (this.isCidLink(key)) {
          dagTree[key] = crushed[key].cid
        } else {
          dagTree[key] = crushed[key]
        }
      } else if (Array.isArray(slice)) {
        const awaits = slice.map((v) => {
          if (v instanceof IpldInterface) {
            return v.crush(resolver)
          }
          return v
        })
        const crushes = await Promise.all(awaits)
        crushed[key] = crushes
        if (this.isCidLink(key)) {
          dagTree[key] = crushes.map((v) => v.cid)
        } else {
          dagTree[key] = crushes
        }
      }
    }
    crushed.#ipldBlock = await encode(dagTree)
    crushed.#crushed = crushed
    crushed.#previous = this.#crushed
    return crushed
  }
  async getDiffBlocks() {
    // diffs since the last time we crushed
    assert(!this.isModified())
    const blocks = new Map()
    const previous = this.#previous
    if (previous === this) {
      return blocks
    }
    blocks.set(this.ipldBlock.cid.toString(), this.ipldBlock)
    for (const key in this) {
      const thisValue = this[key]
      const prevValue = previous && previous[key]
      if (this.isCidLink(key)) {
        if (Array.isArray(thisValue)) {
          // what if the array is an ipldstruct type, included in classMap ?
          // what if it is plain primitive values ?
          // what if I wanted an array of ipld, but not crush ?
          const awaits = thisValue.map(async (v, i) => {
            if (!prevValue || !prevValue[i]) {
              return await v.getDiffBlocks()
            }
            if (!v.cid.equals(prevValue[i].cid)) {
              return await v.getDiffBlocks(prevValue[i])
            }
          })
          const valueBlocks = await Promise.all(awaits)
          valueBlocks.forEach((map) => merge(blocks, map))
        } else {
          assert(prevValue === undefined || prevValue instanceof IpldInterface)
          if (!prevValue || !thisValue.cid.equals(prevValue.cid)) {
            const valueBlocks = await thisValue.getDiffBlocks(prevValue)
            merge(blocks, valueBlocks)
          }
        }
      }
    }
    return blocks
  }
  static async uncrush(rootCid, resolver, options) {
    // throw if resolver does not have what we are looking for
    // recursively rebuild everything, from bottom first ?
    assert(rootCid instanceof CID, `rootCid must be a CID, got ${rootCid}`)
    assert(typeof resolver === 'function', `resolver must be a function`)
    const block = await resolver(rootCid)
    // TODO check the schema
    const map = { ...block.value }
    for (const key in map) {
      if (map[key] instanceof CID) {
        const childClass = this.getClassFor(key)
        map[key] = await childClass.uncrush(map[key], resolver, options)
      }
    }
    const instance = new this()
    Object.assign(instance, map)
    instance.#ipldBlock = block
    instance.#crushed = instance
    instance.#previous = instance
    if (typeof instance.assertLogic === 'function') {
      instance.assertLogic()
    }
    return instance
  }
  setMap(map) {
    assert.strictEqual(typeof map, 'object')
    const inflated = { ...map }
    for (const key in map) {
      assert(map[key] !== undefined, `${key} is undefined`)
      assert(typeof map[key].then !== 'function', `${key} is a promise`)
      if (map[key] === this[key]) {
        delete inflated[key]
        continue
      }
      if (this.constructor.classMap[key]) {
        if (this[key] instanceof IpldStruct) {
          let next = map[key]
          if (!(next instanceof IpldStruct)) {
            next = this[key].setMap(map[key])
          }
          assert(next instanceof this.constructor.classMap[key])
          inflated[key] = next
        }
      }
      // TODO handle arrays
    }
    if (Object.keys(inflated).length === 0) {
      return this
    }
    const next = this.clone()
    next.#ipldBlock = undefined // this is now modified
    Object.assign(next, inflated)
    return next
  }
  delete(key) {
    assert.strictEqual(typeof key, 'string')
    assert(key)
    if (this[key] === undefined) {
      return this
    }
    const next = this.clone()
    delete next[key]
    next.#ipldBlock = undefined // this is now modified
    return next
  }
}

const merge = (dst, src) => {
  assert(dst instanceof Map)
  assert(src instanceof Map)
  for (const [key, value] of src) {
    dst.set(key, value)
  }
}
