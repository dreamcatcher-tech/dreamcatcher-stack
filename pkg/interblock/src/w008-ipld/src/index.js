export { Binary } from './Binary'
export { RequestId } from './RequestId'
export { Request } from './Request'
export { RxRequest } from './RxRequest'
export { Reply } from './Reply'
export { RxReply } from './RxReply'
export { Address } from './Address'
export { PulseLink, HistoricalPulseLink } from './PulseLink'
export { TxQueue } from './TxQueue'
export { Tx } from './Tx'
export { RxQueue } from './RxQueue'
export { Rx } from './Rx'
export { Channel } from './Channel'
export { Config } from './Config'
export { PublicKey } from './PublicKey'
export { Validators } from './Validators'
export { ChannelsHamt } from './ChannelsHamt'
export { ChildrenHamt } from './ChildrenHamt'
export { UplinksHamt } from './UplinksHamt'
export { DownlinksHamt } from './DownlinksHamt'
export { SymlinksHamt } from './SymlinksHamt'
export { AddressesHamt } from './AddressesHamt'
export { Channels } from './Channels'
export { Transmissions } from './Transmissions'
export { Network } from './Network'
export { AsyncRequest } from './AsyncRequest'
export { AsyncTrail } from './AsyncTrail'
export { Pending } from './Pending'
export { State } from './State'
export { Timestamp } from './Timestamp'
export { Dmz } from './Dmz'
export { Provenance } from './Provenance'
export { Interpulse } from './Interpulse'
export { Pulse } from './Pulse'
export { Keypair } from './Keypair'
export { decode } from './IpldUtils'
export { IpldInterface } from './IpldInterface'
export { Hamt } from './Hamt'
