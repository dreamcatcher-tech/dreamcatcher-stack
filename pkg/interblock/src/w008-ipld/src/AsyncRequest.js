import assert from 'assert-fast'
import { Request, RequestId, Reply } from '.'
import { IpldStruct } from './IpldStruct'
import equals from 'fast-deep-equal'

/**
    type AsyncRequest struct {
        request &Request
        to String
        id RequestId
        settled optional &Reply
    }
*/
export class AsyncRequest extends IpldStruct {
  static create(request, to) {
    assert(request instanceof Request)
    assert.strictEqual(typeof to, 'string')
    assert(to)
    const instance = new AsyncRequest()
    instance.request = request
    instance.to = to
    return instance
  }
  setId(id) {
    assert(id instanceof RequestId)
    return this.setMap({ id })
  }
  settle(reply) {
    assert(reply instanceof Reply)
    assert(!this.isSettled())
    return this.setMap({ settled: reply })
  }
  isRequestMatch(request) {
    assert(request instanceof Request)
    return equals(this.request, request)
  }
  isSettled() {
    return this.settled !== undefined
  }
  assertLogic() {
    if (this.settled) {
      assert(this.id !== undefined, `cannot settle unidentified requests`)
    }
  }
}
