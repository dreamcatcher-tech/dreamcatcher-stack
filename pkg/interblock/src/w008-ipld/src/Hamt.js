import equals from 'fast-deep-equal'
import { toString } from 'uint8arrays/to-string'
import Immutable from 'immutable'
import { IpldInterface } from './IpldInterface'
import { IpldStruct } from './IpldStruct'
import { create, load } from 'ipld-hashmap'
import { hasher as blockHasher } from './IpldUtils'
import * as blockCodec from '@ipld/dag-cbor'
import { CID } from 'multiformats/cid'
import assert from 'assert-fast'
import Debug from 'debug'
import { PutStore } from './PutStore'
const debug = Debug('interblock:models:hamt')

const hamtOptions = { blockHasher, blockCodec }
export class Hamt extends IpldInterface {
  #valueClass
  #isMutable
  #putStore
  #hashmap
  #gets = Immutable.Map()
  #sets = Immutable.Map()
  #deletes = Immutable.Set()
  static create(valueClass, isMutable = false) {
    if (valueClass) {
      assert(valueClass.prototype instanceof IpldStruct, 'Not IpldStruct type')
    }
    const instance = new this()
    instance.#valueClass = valueClass
    instance.#isMutable = isMutable
    return instance
  }
  // TODO allow WeakGet, WeakEntries, to only give what is loaded
  #clone() {
    const next = new this.constructor()
    next.#valueClass = this.#valueClass
    next.#isMutable = this.#isMutable
    next.#putStore = this.#putStore && this.#putStore.clone()
    next.#gets = this.#gets
    next.#sets = this.#sets
    next.#deletes = this.#deletes
    if (this.#hashmap) {
      next.#hashmap = new this.#hashmap.constructor(this.#hashmap._iamap)
    }
    return next
  }
  [Symbol.for('nodejs.util.inspect.custom')]() {
    return this.#gets.toJS()
  }
  async set(key, value) {
    assertKey(key)
    if (this.#valueClass) {
      assert(value instanceof this.#valueClass, `Not correct class type`)
    }
    if (!this.#isMutable && (await this.#has(key))) {
      throw new Error(`Cannot overwrite key: ${key}`)
    }
    const next = this.#clone()
    next.#gets = this.#gets.set(key, value)
    next.#sets = this.#sets.set(key, value)
    next.#deletes = this.#deletes.remove(key)
    return next
  }
  delete(key) {
    assertKey(key)
    if (!this.#gets.has(key)) {
      throw new Error(`non existent key: ${key}`)
    }
    const next = this.#clone()
    next.#gets = this.#gets.remove(key)
    next.#sets = this.#sets.remove(key)
    next.#deletes = this.#deletes.add(key)
    return next
  }
  async #has(key) {
    assertKey(key)
    if (this.#gets.has(key)) {
      return true
    }
    if (this.#deletes.has(key)) {
      return false
    }
    if (!this.#hashmap) {
      return false
    }
    return await this.#hashmap.has(key)
  }
  async has(key) {
    return await this.#has(key)
  }
  async get(key) {
    assertKey(key)
    if (!(await this.#has(key))) {
      throw new Error(`key not present: ${key}`)
    }
    if (this.#gets.has(key)) {
      return this.#gets.get(key)
    }
    let value = await this.#hashmap.get(key)
    assert(value !== undefined)
    if (this.#valueClass) {
      assert(value instanceof CID)
      const resolver = (cid) => this.#putStore.getBlock(cid)
      value = await this.#valueClass.uncrush(value, resolver)
    }
    this.#gets = this.#gets.set(key, value)
    return value
  }
  isModified() {
    return !!this.#sets.size || !!this.#deletes.size || !this.#hashmap
  }
  get cid() {
    assert(this.#hashmap)
    assert(!this.isModified())
    return this.#hashmap.cid
  }
  get ipldBlock() {
    throw new Error('Not Implemented')
  }
  get crushedSize() {
    throw new Error('Not Implemented')
  }
  async crush(resolver = () => {}) {
    assert.strictEqual(typeof resolver, 'function')
    const next = this.#clone()
    if (!this.isModified()) {
      return next
    }
    let hashmap = next.#hashmap
    const putStore = new PutStore(resolver, next.#putStore)
    if (!hashmap) {
      hashmap = await create(putStore, hamtOptions)
    } else {
      hashmap = await load(putStore, hashmap.cid, hamtOptions)
    }

    for (const key of next.#deletes) {
      debug(`delete`, key)
      if (!(await hashmap.has(key))) {
        throw new Error(`non existent key: ${key}`)
      }
      await hashmap.delete(key)
    }
    for (const [key, value] of next.#sets) {
      debug('set:', key, value)
      if (!this.#isMutable && (await hashmap.has(key))) {
        throw new Error(`cannot overwrite key: ${key}`)
      }
      if (next.#valueClass) {
        const crushed = await value.crushToCid(resolver)
        next.#gets = next.#gets.set(key, crushed)
        await hashmap.set(key, crushed.cid)
        const diffs = crushed.getDiffBlocks()
        for (const [, ipldBlock] of diffs) {
          if (!putStore.hasBlock(ipldBlock.cid)) {
            putStore.putBlock(ipldBlock.cid, ipldBlock)
          }
        }
      } else {
        await hashmap.set(key, value)
      }
    }
    putStore.trim(hashmap.cid)
    next.#hashmap = hashmap
    next.#sets = next.#sets.clear()
    next.#deletes = next.#deletes.clear()
    next.#putStore = putStore
    return next
  }
  static async uncrush(cid, resolver, valueClass, isMutable) {
    assert(cid instanceof CID, `rootCid must be a CID, got ${cid}`)
    assert(typeof resolver === 'function', `resolver must be a function`)

    const instance = this.create(valueClass, isMutable)
    instance.#putStore = new PutStore(resolver)
    const block = await resolver(cid)
    instance.#putStore.putBlock(cid, block)
    const hashmap = await load(instance.#putStore, cid, hamtOptions)
    instance.#hashmap = hashmap
    return instance
  }
  getDiffBlocks() {
    // This only stores diff since the last call to crush()
    assert(!this.isModified())
    assert(this.#hashmap)
    assert(this.#putStore)

    const { cid } = this
    debug(`get diff for CID`, cid)
    return this.#putStore.getDiffBlocks(cid)
  }
  entries() {
    assert(!this.isModified())
    return this.#hashmap.entries()
  }
  async compare(other) {
    if (!other) {
      other = this.constructor.create(this.#valueClass, this.#isMutable)
    }
    assert(other instanceof this.constructor)
    assert(!this.isModified())
    assert(!other.isModified())

    const links = [{ cid: this.#hashmap.cid, otherCid: other.#hashmap.cid }]
    let limit = 10000
    const mergedDiff = {}
    while (links.length > 0 && limit--) {
      const { cid, otherCid } = links.shift()
      const value = await safelyGetBlock(this.#putStore, cid)
      const otherValue = await safelyGetBlock(other.#putStore, otherCid)

      const [, data] = value
      const [, otherData] = otherValue
      const max = Math.max(data.length, otherData.length)
      for (let i = 0; i < max; i++) {
        const element = data[i]
        const otherElement = otherData[i]
        if (CID.asCID(element) || CID.asCID(otherElement)) {
          const isBothCid = CID.asCID(element) && CID.asCID(otherElement)
          if (isBothCid) {
            if (!CID.asCID(element).equals(otherElement)) {
              debug('different links found')
              links.push({ cid: element, otherCid: otherElement })
            }
          } else {
            debug('one link is not a CID', element, otherElement)
            links.push({
              cid: CID.asCID(element),
              otherCid: CID.asCID(otherElement),
            })
            patchBuckets(element, otherElement, mergedDiff)
          }
          // if cid vs bucket, compare bucket against blank, then cid against blank
        } else {
          patchBuckets(element, otherElement, mergedDiff)
        }
      }
    }

    return mergedDiff
  }
}
const safelyGetBlock = async (putStore, cid) => {
  if (!cid) {
    return [undefined, []]
  }
  const block = await putStore.getBlock(cid)
  assert(block)
  let { value } = block
  value = value.hamt ? value.hamt : value
  return value
}
const assertKey = (key) => {
  assert(key !== undefined)
  assert(key !== '')
  assert(typeof key === 'string' || Number.isInteger(key))
}
const patchBuckets = (bucket, otherBucket, mergedDiff) => {
  bucket = Array.isArray(bucket) ? bucket : []
  otherBucket = Array.isArray(otherBucket) ? otherBucket : []

  const patch = compareBuckets(bucket, otherBucket)
  if (patch) {
    debug('node was modified', patch)
    merge(mergedDiff, patch)
  }
}
const compareBuckets = (bucket, otherBucket) => {
  const diff = { added: [], deleted: [], modified: [] }
  const map = bucketToMap(bucket)
  const otherMap = bucketToMap(otherBucket)

  otherMap.forEach((value, key) => {
    if (!map.has(key)) {
      diff.deleted.push(key)
    }
  })
  map.forEach((value, key) => {
    if (!otherMap.has(key)) {
      diff.added.push(key)
    } else {
      const otherValue = otherMap.get(key)
      if (!equals(value, otherValue)) {
        diff.modified.push(key)
      }
    }
  })
  if (diff.added.length || diff.deleted.length || diff.modified.length) {
    return diff
  }
}
const bucketToMap = (bucket) => {
  const map = new Map()
  if (!bucket) {
    return map
  }
  for (const [key, value] of bucket) {
    map.set(toString(key), value)
  }
  return map
}
const merge = (diff, patch) => {
  diff.added = diff.added || new Set()
  diff.deleted = diff.deleted || new Set()
  diff.modified = diff.modified || new Set()

  patch.added.forEach((key) => {
    if (diff.deleted.has(key)) {
      diff.deleted.delete(key)
      return
    }
    diff.added.add(key)
  })
  patch.deleted.forEach((key) => {
    if (diff.added.has(key)) {
      diff.added.delete(key)
      return
    }
    diff.deleted.add(key)
  })
  patch.modified.forEach((key) => {
    diff.modified.add(key)
  })
}
