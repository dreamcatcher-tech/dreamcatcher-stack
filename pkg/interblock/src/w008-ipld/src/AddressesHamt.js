import { Hamt } from './Hamt'
import { Address } from '.'
import assert from 'assert-fast'
export class AddressesHamt extends Hamt {
  get isBakeSkippable() {
    return true
  }
  async set(address, channelId) {
    assert(address instanceof Address)
    assert(Number.isInteger(channelId))
    assert(channelId >= 0)
    assert(address.isRemote())
    const key = address.getChainId()
    return await super.set(key, channelId)
  }
  async has(address) {
    assert(address instanceof Address)
    assert(address.isRemote())
    return await super.has(address.getChainId())
  }
  async delete(address) {
    assert(await this.has(address), `address not present ${address}`)
    return await super.delete(address.getChainId())
  }
  async get(address) {
    assert(address instanceof Address)
    return await super.get(address.getChainId())
  }
  async compare(other) {
    let { added, deleted, modified } = await super.compare(other)
    added = new Set([...added].map((key) => Address.fromChainId(key)))
    deleted = new Set([...deleted].map((key) => Address.fromChainId(key)))
    modified = new Set([...modified].map((key) => Address.fromChainId(key)))
    return { added, deleted, modified }
  }
}
