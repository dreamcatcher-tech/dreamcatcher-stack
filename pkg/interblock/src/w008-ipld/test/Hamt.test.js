import { IpldStruct } from '../src/IpldStruct'
import { assert } from 'chai/index.mjs'
import { Hamt } from '../src/Hamt'
import Debug from 'debug'
import random from 'random'

const debug = Debug('tests')

describe('Hamt', () => {
  test('basic', async () => {
    const base = Hamt.create()
    assert(base.isModified())
    let hamt = await base.set('testkey', 'testvalue')
    assert(hamt.isModified())
    assert(base !== hamt)
    assert.strictEqual(await hamt.get('testkey'), 'testvalue')
    const crushed = await hamt.crush()
    const diffs = crushed.getDiffBlocks()
    assert.strictEqual(diffs.size, 1)
    const resolver = (cid) => diffs.get(cid.toString())
    const uncrushed = await Hamt.uncrush(crushed.cid, resolver)
    assert(await uncrushed.has('testkey'))
    const result = await uncrushed.get('testkey')
    assert.strictEqual(result, 'testvalue')
  })
  test('multiple keys', async () => {
    let hamt = Hamt.create()
    hamt = await hamt.crush()
    const diffs = hamt.getDiffBlocks()
    assert.strictEqual(diffs.size, 1)
    expect(diffs).toMatchSnapshot()
    const resolver = (cid) => diffs.get(cid.toString())
    hamt = await Hamt.uncrush(hamt.cid, resolver)
    for (let i = 0; i < 100; i++) {
      hamt = await hamt.set('test-' + i, { test: 'test-' + i })
    }
    let start = Date.now()
    hamt = await hamt.crush()
    debug(`crush: ${Date.now() - start} ms`)
    start = Date.now()
    const bigDiffs = hamt.getDiffBlocks()
    debug(`diffs: ${Date.now() - start} ms`)
    start = Date.now()
    const bigResolver = (cid) => bigDiffs.get(cid.toString())
    hamt = await Hamt.uncrush(hamt.cid, bigResolver)
    debug(`uncrush: ${Date.now() - start} ms`)

    start = Date.now()
    assert.deepEqual(await hamt.get('test-59'), { test: 'test-59' })
    debug(`deep get: ${Date.now() - start} ms`)

    start = Date.now()
    assert.deepEqual(await hamt.get('test-58'), { test: 'test-58' })
    debug(`deep get 2: ${Date.now() - start} ms`)
  })
  test('with class', async () => {
    class TestClass extends IpldStruct {}
    let hamt = Hamt.create(TestClass)
    const msg = 'Not correct class type'
    await expect(hamt.set('test', { a: 'b' })).rejects.toThrow(msg)
    const testInstance = new TestClass()
    hamt = await hamt.set('test', testInstance)
    assert((await hamt.get('test')) instanceof TestClass)
    hamt = await hamt.crush()
    const diffs = hamt.getDiffBlocks()
    expect(diffs.size).toBe(2)
    expect(diffs).toMatchSnapshot()
    const resolver = (cid) => diffs.get(cid.toString())
    hamt = await Hamt.uncrush(hamt.cid, resolver, TestClass)
    const result = await hamt.get('test')
    assert(result instanceof TestClass)
  })
  test('throws on existing key', async () => {
    let hamt = Hamt.create()
    await expect(hamt.get('bogus key')).rejects.toThrow('bogus key')
    hamt = await hamt.set('some key', 'some value')
    assert.strictEqual(await hamt.get('some key'), 'some value')
    await expect(hamt.set('some key', 'over')).rejects.toThrow(
      'Cannot overwrite'
    )
  })
  test.only('compare blank', async () => {
    const size = 53
    const raw = await hamtFactory(size)
    const base = await raw.crush()
    const diff = await base.compare()
    expect(diff.added.size).toEqual(53)
    expect(diff.deleted.size).toEqual(0)
    expect(diff.modified.size).toEqual(0)
  })
  test('diffing', async () => {
    const raw = await hamtFactory()
    const base = await raw.crush()
    debug('start')

    let added = await base.set('addedKey', { test: 'added' })
    added = await added.crush()
    const addDiff = await added.compare(base)
    debug(addDiff)
    expect([...addDiff.added]).toEqual(['addedKey'])
    expect(addDiff.modified.size).toEqual(0)
    expect(addDiff.deleted.size).toEqual(0)

    let deleted = await base.delete('test-0')
    deleted = await deleted.crush()
    const delDiff = await deleted.compare(base)
    debug(delDiff)
    expect([...delDiff.deleted]).toEqual(['test-0'])
    expect(delDiff.modified.size).toEqual(0)
    expect(delDiff.added.size).toEqual(0)

    let modified = await base.set('test-1', { test: 'modified' })
    modified = await modified.crush()
    const modDiff = await modified.compare(base)
    debug(modDiff)
    expect([...modDiff.modified]).toEqual(['test-1'])
    expect(modDiff.deleted.size).toEqual(0)
    expect(modDiff.added.size).toEqual(0)
  })
  test('diff stress test', async () => {
    for (let i = 0; i < 50; i++) {
      const ipfsPersistence = new Map()
      const size = random.int(10, 100)
      debug(`iteration %i with size %i`, i, size)
      const diff = { added: new Set(), modified: new Set(), deleted: new Set() }

      let base = await hamtFactory(size)
      base = await crush(base, ipfsPersistence)
      let next = base
      // choose an operation at random
      if (random.boolean()) {
        const addCount = random.int(1, 50)
        next = await add(next, diff, addCount)
        next = await crush(next, ipfsPersistence)
      }
      if (random.boolean()) {
        const modCount = random.int(10, 50)
        next = await mod(next, diff, modCount)
        next = await crush(next, ipfsPersistence)
      }
      if (random.boolean()) {
        const delCount = random.int(1, 50)
        next = await del(next, diff, delCount)
        next = await crush(next, ipfsPersistence)
      }

      debug('crushing')
      next = await next.crush()
      debug('comparing')
      const nextDiff = await next.compare(base)
      debug('comparing done')
      const addDiff = symmetricDifference(diff.added, nextDiff.added)
      expect(addDiff.size).toEqual(0)
      const delDiff = symmetricDifference(diff.deleted, nextDiff.deleted)
      expect(nextDiff.deleted).toEqual(diff.deleted)
      expect(delDiff.size).toEqual(0)
      const modDiff = symmetricDifference(diff.modified, nextDiff.modified)
      expect(modDiff.size).toEqual(0)
    }
  })
  test.todo('recursive crush')
})
const crush = async (hamt, ipfsMap) => {
  debug('crushing')
  const resolver = (cid) => ipfsMap.get(cid.toString())
  hamt = await hamt.crush(resolver)
  const diffBlocks = hamt.getDiffBlocks()
  diffBlocks.forEach((v, k) => ipfsMap.set(k, v))
  return hamt
}
const mod = async (hamt, diff, modCount) => {
  debug(`modifying %i`, modCount)
  const keys = []
  for await (const [key] of hamt.entries()) {
    keys.push(key)
  }
  const shuffled = keys
    .map((value) => ({ value, sort: Math.random() }))
    .sort((a, b) => a.sort - b.sort)
    .map(({ value }) => value)
  for (let i = 0; i < modCount; i++) {
    if (!shuffled.length) {
      break
    }
    const key = shuffled.pop()
    const value = await hamt.get(key)
    hamt = await hamt.set(key, { ...value, modified: i })
    if (!diff.added.has(key)) {
      diff.modified.add(key)
    }
  }
  return hamt
}
const del = async (hamt, diff, delCount) => {
  debug(`deleting %i`, delCount)
  const keys = []
  for await (const [key] of hamt.entries()) {
    keys.push(key)
  }
  const shuffled = keys
    .map((value) => ({ value, sort: Math.random() }))
    .sort((a, b) => a.sort - b.sort)
    .map(({ value }) => value)
  debug('keys listed')
  for (let i = 0; i < delCount; i++) {
    if (!shuffled.length) {
      break
    }
    const key = shuffled.pop()
    hamt = await hamt.delete(key)
    if (diff.added.has(key)) {
      diff.added.delete(key)
    } else {
      if (diff.modified.has(key)) {
        diff.modified.delete(key)
      }
      diff.deleted.add(key)
    }
  }
  return hamt
}
const add = async (hamt, diff, addCount) => {
  debug(`adding %i`, addCount)
  for (let i = 0; i < addCount; i++) {
    const key = 'added-' + i
    assert(!(await hamt.has(key)), 'key already exists')
    hamt = await hamt.set(key, { test: key })
    diff.added.add(key)
    assert(!diff.deleted.has(key))
    assert(!diff.modified.has(key))
  }
  return hamt
}
const hamtFactory = async (count = 100) => {
  const valueClass = undefined
  const isMutable = true
  let hamt = Hamt.create(valueClass, isMutable)
  for (let i = 0; i < count; i++) {
    hamt = await hamt.set('test-' + i, { test: 'test-' + i })
  }
  return hamt
}
const symmetricDifference = (setA, setB) => {
  const _difference = new Set(setA)
  for (const elem of setB) {
    if (_difference.has(elem)) {
      _difference.delete(elem)
    } else {
      _difference.add(elem)
    }
  }
  return _difference
}
