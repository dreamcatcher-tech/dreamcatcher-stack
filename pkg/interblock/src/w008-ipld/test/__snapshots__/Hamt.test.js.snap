// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hamt multiple keys 1`] = `
Map {
  "bafyreig3w5cuffzshczi5xzwnp4igna5wehxcisr53jcjtrfxcnbgzwrui" => Block {
    "asBlock": [Circular],
    "bytes": Uint8Array [
      163,
      100,
      104,
      97,
      109,
      116,
      130,
      68,
      0,
      0,
      0,
      0,
      128,
      103,
      104,
      97,
      115,
      104,
      65,
      108,
      103,
      18,
      106,
      98,
      117,
      99,
      107,
      101,
      116,
      83,
      105,
      122,
      101,
      3,
    ],
    "cid": {
      "/": "bafyreig3w5cuffzshczi5xzwnp4igna5wehxcisr53jcjtrfxcnbgzwrui",
    },
    "value": {
      "bucketSize": 3,
      "hamt": [
        Uint8Array [
          0,
          0,
          0,
          0,
        ],
        [],
      ],
      "hashAlg": 18,
    },
  },
}
`;

exports[`Hamt with class 1`] = `
Map {
  "bafyreibrhfhhayfcwgm3gqdokiffu2rczo6diswuhgmnt5yf2thjeawu44" => Block {
    "asBlock": [Circular],
    "bytes": Uint8Array [
      163,
      100,
      104,
      97,
      109,
      116,
      130,
      68,
      0,
      0,
      8,
      0,
      129,
      129,
      130,
      68,
      116,
      101,
      115,
      116,
      216,
      42,
      88,
      37,
      0,
      1,
      113,
      18,
      32,
      193,
      154,
      121,
      127,
      161,
      253,
      89,
      12,
      210,
      229,
      180,
      45,
      28,
      245,
      242,
      70,
      226,
      155,
      145,
      104,
      78,
      47,
      135,
      64,
      75,
      129,
      220,
      52,
      92,
      122,
      86,
      160,
      103,
      104,
      97,
      115,
      104,
      65,
      108,
      103,
      18,
      106,
      98,
      117,
      99,
      107,
      101,
      116,
      83,
      105,
      122,
      101,
      3,
    ],
    "cid": {
      "/": "bafyreibrhfhhayfcwgm3gqdokiffu2rczo6diswuhgmnt5yf2thjeawu44",
    },
    "value": {
      "bucketSize": 3,
      "hamt": [
        Uint8Array [
          0,
          0,
          8,
          0,
        ],
        [
          [
            [
              Uint8Array [
                116,
                101,
                115,
                116,
              ],
              {
                "code": 113,
                "hash": Uint8Array [
                  18,
                  32,
                  193,
                  154,
                  121,
                  127,
                  161,
                  253,
                  89,
                  12,
                  210,
                  229,
                  180,
                  45,
                  28,
                  245,
                  242,
                  70,
                  226,
                  155,
                  145,
                  104,
                  78,
                  47,
                  135,
                  64,
                  75,
                  129,
                  220,
                  52,
                  92,
                  122,
                  86,
                  160,
                ],
                "version": 1,
              },
            ],
          ],
        ],
      ],
      "hashAlg": 18,
    },
  },
  "bafyreigbtj4x7ip5legnfznufuopl4sg4knzc2cof6duas4b3q2fy6swua" => Block {
    "asBlock": [Circular],
    "bytes": Uint8Array [
      160,
    ],
    "cid": {
      "code": 113,
      "hash": Uint8Array [
        18,
        32,
        193,
        154,
        121,
        127,
        161,
        253,
        89,
        12,
        210,
        229,
        180,
        45,
        28,
        245,
        242,
        70,
        226,
        155,
        145,
        104,
        78,
        47,
        135,
        64,
        75,
        129,
        220,
        52,
        92,
        122,
        86,
        160,
      ],
      "version": 1,
    },
    "value": {},
  },
}
`;
